core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala

```diff
@@ -592,16 +593,75 @@ class RemoteIndexCacheTest {
     verifyFetchIndexInvocation(count = 1)
   }
 
+  @Test
+  def testConcurrentRemoveReadForCache(): Unit = {
+    // Create a spy Cache Entry
+    val rlsMetadata = new RemoteLogSegmentMetadata(RemoteLogSegmentId.generateNew(idPartition), baseOffset, lastOffset,
+      time.milliseconds(), brokerId, time.milliseconds(), segmentSize, Collections.singletonMap(0, 0L))
+
+    val timeIndex = spy(createTimeIndexForSegmentMetadata(rlsMetadata, new File(tpDir, DIR_NAME)))
+    val txIndex = spy(createTxIndexForSegmentMetadata(rlsMetadata, new File(tpDir, DIR_NAME)))
+    val offsetIndex = spy(createOffsetIndexForSegmentMetadata(rlsMetadata, new File(tpDir, DIR_NAME)))
+
+    val spyEntry = spy(new RemoteIndexCache.Entry(offsetIndex, timeIndex, txIndex))
+    cache.internalCache.put(rlsMetadata.remoteLogSegmentId().id(), spyEntry)
+
+    assertCacheSize(1)
+
+    var entry: RemoteIndexCache.Entry = null
+
+    val latchForCacheRead = new CountDownLatch(1)
+    val latchForCacheRemove = new CountDownLatch(1)
+    val latchForTestWait = new CountDownLatch(1)
+
+    doAnswer((invocation: InvocationOnMock) => {
+      // Signal the CacheRead to unblock itself
+      latchForCacheRead.countDown()
+      // Wait for signal to start renaming the files
+      latchForCacheRemove.await()
+      // Calling the markForCleanup() actual method to start renaming the files
+      invocation.callRealMethod()
```

Comment:
> 3. readCache executed and finished because no lock is pending on the remove operation

No, it's not right. The lock is in the concurrentMap inside cache. 
But checking it again, I found the lock is when "updating the entry". That is, we have 2 threads trying to do:
1. create a new entry for the key
2. remove the entry for the key
We can't make sure which one will complete first. All we can make sure is the 2 operations are atomic. So, after L646:
```
      // Wait for signal to complete the test
      latchForTestWait.await()
    
     // Here, we can't make sure if the cache size is 0 or 1 at this point. I think we can only do is to make sure either it's cache size 1 and file existed (create goes later), or cache size is 0 and file inexisted (remove goes later)

    // So, maybe we verify with this:
      if (Files.exists(entry.offsetIndex().file().toPath)) {
          assertCacheSize(1)
     } else {
        assertCacheSize(0)
     } 
```

WDYT?

Reply:
@showuon Yes seems correct , as we do not have much control which will get executed we can assert Cache size.  

As there may be a case where `remove` and `read` operation  has a bug , and nothing happens , but this test case would still pass as  file still exists. Adding one more validation when file exists would ensure the correctness of the test case. 
```
if (Files.exists(entry.offsetIndex().file().toPath)) {
          assertCacheSize(1)
     //  validate the timestamp of the file created before running the `remove`  and  `read` operation concurrently with the file recreated after the operation and it should be lesser than the later.  
```

Reply:
Thank you @showuon @iit2009060 for your valuable comments. I will update the PR with the changes.

Reply:
Updated test case. 
@iit2009060 I test the test case without the fix and test case is failing as expected so I haven't added timestamp. Can you please tell me in which particular case the test case could pass with the older code (without fix).

Reply:
Also I found below two test cases flaky which doesn't seems to be related to this change.
- testIndexFileAlreadyExistOnDiskButNotInCache
- testCorrectnessForCacheAndIndexFilesWhenResizeCache

Reply:
`testIndexFileAlreadyExistOnDiskButNotInCache` is failing locally sometimes with the same failure mode as `testCorrectnessForCacheAndIndexFilesWhenResizeCache`